File: ruby/lib/factory.rb

module Coul
  module Factory

    def self.build_pong
      header + "PONG\n\n"
    end

    def self.build_ping
      header + "PING\n\n"
    end

    def self.build_smsg(nick, server, channel, timestamp, message)
      "#{header}SMSG #{nick}@#{server} #{channel} #{timestamp}\n#{message}\n"
    end

    def self.build_msg(channel, message)
      "#{header}MSG #{channel}\n#{message}\n\n"
    end

    def self.build_alert(source, timestamp, message)
      if source == :server
        source = "SERVER"
      elsif source == :network
        source = "NETWORK"
      else
        raise ArgumentError, "source needs to be :server or :network"
      end
      "#{header}ALERT #{source} #{timestamp}\n#{message}\n"
    end


    private

    def self.header
      "COUL " + VERSION + " "
    end

  end
end

File: node/lib/factory.js
var version = require('./version')

module.exports = factory = (function() {
  function header() {
    return "COUL " + version + " "
  }

  function process_body(body) {
    if (body.slice(-1) != "\n") {
      return body + "\n"
    } else {
      return body;
    }
  }

  this.ping = function() {
    return header() + "PING\n\n"
  }

  this.pong = function() {
    return header() + "PONG\n\n"
  }

  this.smsg = function(nick, server, channel, timestamp, message) {
    message = process_body(message)
    return header() + "SMSG " + nick + "@" + server + " " + channel + " " + timestamp + "\n" + message + "\n"
  }

  this.msg = function(channel, message) {
    message = process_body(message)
    return header() + "MSG " + channel + "\n" + message + "\n"
  }

  this.alert = function(source, timestamp, message) {
    if (source != "SERVER" && source != "NETWORK") {
      throw new Error("Source needs to be SERVER or NETWORK")
    }
    message = process_body(message)
    return header() + "ALERT " + source + " " + timestamp + "\n" + message + "\n"
  }

  this.now = {
  }

  this.now.smsg = function(nick, server, channel, message) {
    var now = new Date() / 1000;
    return factory.smsg(nick, server, channel, now, message);
  }

  this.now.alert = function(source, message) {
    var now = new Date() / 1000;
    return factory.alert(source, now, message);
  }

  return this;

})()
